plugins/link/SharedMemory.cpp

```diff
@@ -146,11 +150,47 @@ void *SharedMemory::mapMemory(const char *name, std::size_t size) {
 	m_name.assign(name);
 #endif
 
-	m_size = size;
-
 	if (created) {
-		std::memset(m_data, 0, m_size);
+		fillWithZero();
+	}
+
+	return true;
+}
+
+bool SharedMemory::isMemoryMapped() {
+	return m_data != nullptr;
+}
+
+void SharedMemory::read(LinkedMem *dest) {
+	std::atomic< Chunk > *data = (std::atomic< Chunk > *) m_data;
+
+	Chunk buff[chunkCount];
+
+	for (std::size_t i = 0; i < chunkCount; i++) {
+		buff[i] = data[i].load(std::memory_order_relaxed);
+	}
+
+	// bitcast the chunks with memcpy to avoid running afoul
+	// of type-based alias analysis
+	std::memcpy(dest, buff, sizeof(LinkedMem));
```

Comment:
Could we not have done something like
```cpp
Chunk *buff = reinterpret_cast< Chunk * >(&dest)`
```
before the read-out loop and then save on the additional memory copying?

Reply:
Because of type-based alias analysis, AKA the strict aliasing rule, C and C++ compilers are allowed to assume pointers to types that aren't "similar" don't point to overlapping memory. In C, you can get around this by using unions, but in C++, I don't think there's a standardized way to do type punning like this without a copy.

Reply:
Ouch - another lovely crafted trap luring unwary programs into UB...
Aliasing as bytes would be valid, but we want to use 32-bit ints, so yeah. It seems like we do indeed need a copy.